cmd/clusterctl/internal/util/objs.go

```diff
@@ -105,42 +120,87 @@ func FixImages(objs []unstructured.Unstructured, alterImageFunc func(image strin
 		}
+		objs[i] = *o
+	}
+	return objs, nil
+}
 
-		// Alter the image
-		for j := range d.Spec.Template.Spec.Containers {
-			container := d.Spec.Template.Spec.Containers[j]
-			image, err := alterImageFunc(container.Image)
-			if err != nil {
-				return nil, errors.Wrapf(err, "failed to fix image for container %s in deployment %s", container.Name, d.Name)
-			}
-			container.Image = image
-			d.Spec.Template.Spec.Containers[j] = container
-		}
+func fixDeploymentImages(o *unstructured.Unstructured, alterImageFunc func(image string) (string, error)) error {
+	if o.GetKind() != deploymentKind {
+		return nil
+	}
 
-		for j := range d.Spec.Template.Spec.InitContainers {
-			container := d.Spec.Template.Spec.InitContainers[j]
-			image, err := alterImageFunc(container.Image)
-			if err != nil {
-				return nil, errors.Wrapf(err, "failed to fix image for init container %s in deployment %s", container.Name, d.Name)
-			}
-			container.Image = image
-			d.Spec.Template.Spec.InitContainers[j] = container
-		}
+	// Convert Unstructured into a typed object
+	d := &appsv1.Deployment{}
+	if err := scheme.Scheme.Convert(o, d, nil); err != nil {
+		return err
+	}
 
-		// Convert typed object back to Unstructured
-		if err := scheme.Scheme.Convert(d, o, nil); err != nil {
-			return nil, err
+	var err error
+	d.Spec.Template.Spec, err = fixPodSpecImages(d.Spec.Template.Spec, alterImageFunc)
+	if err != nil {
+		return errors.Wrapf(err, "failed to fix containers in deployment %s", d.Name)
+	}
+
+	// Convert typed object back to Unstructured
+	if err := scheme.Scheme.Convert(d, o, nil); err != nil {
+		return err
+	}
+	return nil
+}
+
+func fixDaemonSetImages(o *unstructured.Unstructured, alterImageFunc func(image string) (string, error)) error {
+	if o.GetKind() != daemonSetKind {
+		return nil
+	}
+
+	// Convert Unstructured into a typed object
+	d := &appsv1.DaemonSet{}
+	if err := scheme.Scheme.Convert(o, d, nil); err != nil {
+		return err
+	}
+
+	var err error
+	d.Spec.Template.Spec, err = fixPodSpecImages(d.Spec.Template.Spec, alterImageFunc)
+	if err != nil {
+		return errors.Wrapf(err, "failed to fix containers in deamonSet %s", d.Name)
+	}
+	// Convert typed object back to Unstructured
+	if err := scheme.Scheme.Convert(d, o, nil); err != nil {
+		return err
+	}
+	return nil
+}
+
+func fixPodSpecImages(podSpec v1.PodSpec, alterImageFunc func(image string) (string, error)) (v1.PodSpec, error) {
+	var err error
+	podSpec.Containers, err = fixContainersImage(podSpec.Containers, alterImageFunc)
+	if err != nil {
+		return podSpec, errors.Wrapf(err, "failed to fix containers")
+	}
+	podSpec.InitContainers, err = fixContainersImage(podSpec.InitContainers, alterImageFunc)
+	if err != nil {
+		return podSpec, errors.Wrapf(err, "failed to fix init containers")
+	}
+	return podSpec, nil
+}
+
+func fixContainersImage(containers []v1.Container, alterImageFunc func(image string) (string, error)) ([]v1.Container, error) {
+	for j := range containers {
```

Comment:
```suggestion
func fixContainersImage(containers []*v1.Container, alterImageFunc func(image string) (string, error)) ([]v1.Container, error) {
	for j := range containers {
```

Any reason this isn't a pointer?

Reply:
If I do this, then I need to map containers with pointers here, and I don't like it too much

```
	podSpec.Containers, err = fixContainersImage(podSpec.Containers, alterImageFunc)
	if err != nil {
		return podSpec, errors.Wrapf(err, "failed to fix containers")
	}
	podSpec.InitContainers, err = fixContainersImage(podSpec.InitContainers, alterImageFunc)
	if err != nil {
		return podSpec, errors.Wrapf(err, "failed to fix init containers")
	}
```

In general I like to pass data by value if it's not required

Anyway since it's and array I do not need to pass back the value

Reply:
I've refactored this to avoid passing back data, what do you think?

Reply:
Looks great now, thank you!

Reply:
> @ludusrusso: The following test **failed**, say `/retest` to rerun all failed tests:
> 
> Test name	Commit	Details	Rerun command
> pull-cluster-api-test-main	[8481b9f](https://github.com/kubernetes-sigs/cluster-api/commit/8481b9fda3298d6eddb15e504ae9383ddc86ebd6)	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api/4455/pull-cluster-api-test-main/1382012990940778496)	`/test pull-cluster-api-test-main`
> [Full PR test history](https://prow.k8s.io/pr-history?org=kubernetes-sigs&repo=cluster-api&pr=4455). [Your PR dashboard](https://prow.k8s.io/pr?query=is%3Apr%20state%3Aopen%20author%3Aludusrusso). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/kubernetes-sigs/cluster-api/issues?q=is:issue+is:open) when you hit one in your PR.

@vincepri is this the flaky test or I've broke something?